### PR TITLE
[SPARK-28143][SQL] Expressions without proper constructors should throw AnalysisException

### PR DESCRIPTION
**Pull request by yaooqinn**
## What changes were proposed in this pull request?

SQL queries like `select 1 where in ()`, which miss the attribute name should throw `AnalysisException` with `Invalid number of arguments for function in` as debug message rather than `java.lang.UnsupportedOperationException: empty.init`.

This problem I guess is related to this pr - https://github.com/apache/spark/pull/21226

## How was this patch tested?

1. add ut
2. Manually

**Before**

```scala
select 1 where in ();
19/06/24 13:37:33 ERROR SparkSQLDriver: Failed in [select 1 where in ()]
java.lang.UnsupportedOperationException: empty.init
```

**After**

```
Invalid number of arguments for function in. Expected: 2; Found: 0
```

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 1]
```diff
@@ -581,7 +581,6 @@ object FunctionRegistry {
         val params = Seq.fill(expressions.size)(classOf[Expression])
         val f = constructors.find(_.getParameterTypes.toSeq == params).getOrElse {
           val validParametersCount = constructors
-            .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: This partially reverts #21226. Previous logic was to detect SQL related expressions only but now it includes all constructors that are not related to SQL expressions. Can you elaborate why we revert?

Looks we should rather specially handle `In`'s constructors or handle when `validParametersCount` is empty.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: SQL related expressions has paras which are not just `Expression`s
This logic was added in #21226, I have no idea why this should add in that pr. Seems not related to its topic.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Constructors with expressions only are recognised in SQL.so if a constructor has non-`Expression` argument, it should be excluded.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Constructors may take 
```
1 Expression*; 
2 Seq[Expression]; 
3 (Expression, Expression*);
4 (Expression, Seq[Expression]);
```
as args

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Yea, then can you consider them except other parameters rather than reverting the filter logic entirely?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 2]
```diff
@@ -581,7 +581,9 @@ object FunctionRegistry {
...
+            .filter(_.getParameterTypes.forall { t =>
+              t == classOf[Expression] || t == classOf[Seq[Expression]]
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I think generic type can't be checked runtime. Also, this code path is only for `Expression`. See `f.newInstance(expressions : _*)`.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -106,6 +106,12 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     checkKeywordsExist(sql("describe functioN abcadf"), "Function: abcadf not found.")
   }
 
+  test("SPARK-28143: IN expression missing attribute should throw analysis exception") {
+    val query = "select 1 from where in ()"
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: BTW, is it correct syntax? Looks like we never supported `in` as function before.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 1 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: I think we don't support `4 (Expression, Seq[Expression]);` case here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 4]
```diff
@@ -581,7 +581,9 @@ object FunctionRegistry {
...
+            })
             .map(_.getParameterCount).distinct.sorted
           val expectedNumberOfParameters = if (validParametersCount.length == 1) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Can you rather do this for now?

```scala
if (validParametersCount.length == 0) {
  throw new AnalysisException("'$name' is an invalid function.")
}
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (yeshengm) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: It's a parser thing. The parser has no idea whether in is a keyword or function name.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 2 above)
Comment (yeshengm) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Second that.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: `select 1 from where from in ();` 
:-)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: but actually `in` is a valid function

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: It is kinda weird, I use this query just because it is simple.

a real world wrong query maybe like this

```sql
select dt, os, userid, usertype
    from tbl1
    where dt=date_sub('2019-06-30',1)
    and source='client'
    and in ('iphone','android')  --- here
```

**Comment by cloud-fan** [comment]
does the problem still exist when we turn on `spark.sql.parser.ansi.enabled`?

**Comment by yaooqinn** [comment]
@cloud-fan spark.sql.parser.ansi.enabled=true  works fine
```
spark-sql> select 1 where not in ('2019-06-21');
Error in query:
no viable alternative at input 'in'(line 1, pos 19)

== SQL ==
select 1 where not in ('2019-06-21')
```

**Comment by cloud-fan** [comment]
Then I think it's a parser issue(and is already fixed), we shouldn't hack `FunctionRegistry` for it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 5]
```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
           val validParametersCount = constructors
             .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
             .map(_.getParameterCount).distinct.sorted
-          val expectedNumberOfParameters = if (validParametersCount.length == 1) {
+          val expectedNumberOfParameters = if (validParametersCount.length == 0) {
+            throw new AnalysisException(s"$name is an invalid function.")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Yea, I agree with https://github.com/apache/spark/pull/24947#issuecomment-505260696. Let's just raise an exception like this. It's an invalid function if there's no constructor to call anyway.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 6]
```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
...
+          } else if (validParametersCount.length == 1) {
             validParametersCount.head.toString
           } else {
             validParametersCount.init.mkString("one of ", ", ", " and ") +
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: can we just handle empty `validParametersCount` here? I think it's better to have a unified error message:
```
Invalid number of arguments for function $name. Expected .....
```

**Comment by yaooqinn** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 7]
```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
           val validParametersCount = constructors
             .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
             .map(_.getParameterCount).distinct.sorted
-          val expectedNumberOfParameters = if (validParametersCount.length == 1) {
+          val expectedNumberOfParameters = if (validParametersCount.isEmpty) {
+            constructors.headOption.map(_.getParameterCount).getOrElse(0).toString
+          } else if (validParametersCount.length == 1) {
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: @yaooqinn, can you show the error message before and after this change, and fix the PR description and title? I think this PR now targets throw a better exception for the expression that has no constructors to call.

**Comment by srowen** [comment]
@yaooqinn I think there are some last comments to address?

**Comment by yaooqinn** [comment]
@srowen I have updated the pr description and title as last comment required. If anything I misread, please correct me. Thanks very much.

**Comment by srowen** [comment]
@HyukjinKwon is that OK by you?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 8]
```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
           val validParametersCount = constructors
             .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
             .map(_.getParameterCount).distinct.sorted
-          val expectedNumberOfParameters = if (validParametersCount.length == 1) {
+          val expectedNumberOfParameters = if (validParametersCount.isEmpty) {
+            constructors.headOption.map(_.getParameterCount).getOrElse(0).toString
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: use headOption? This is hacky. We should follow the same logic what we are doing here.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala [hunk 9]
```diff
@@ -583,7 +583,9 @@ object FunctionRegistry {
           val validParametersCount = constructors
             .filter(_.getParameterTypes.forall(_ == classOf[Expression]))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: If we remove this line, what is the outcome?

In this PR, the logic is to add a special case. When unable to find a constructor whose parameters are all Expressions, we use the remaining constructors. [Note, headOption means we only choose the first one] Could you please check whether we should list all of them in the error messages?

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: If we remove this line, it works fine. I did this in the first commit of the pr,  discussed with HyukjinKwon [here](https://github.com/apache/spark/pull/24947/files/4af96a1b90ceb54c64c798215d2f5eefe87ad896#r296583009)

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: No, It doesn't work fine.
We should only recognise the constructor with expressions because those are only able to be used in SQL. Otherwise, it shows incorrect information about possible argument combination which partially reverts #21226.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 8 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: The problem is that we have function registered `in` but that's unable/not supposed to to call via `in(...)`. Possibly there might be more cases like this. Ideally we might have to identify and remove those expressions out but wanted to narrow the scope this PR by such change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala (hunk 9 above)
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: The problem is that we have function registered `in` but that's unable/not supposed to to call via `in(...)` currently. Possibly there might be more cases like this.

**Comment by yaooqinn** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Seems we're blocked by https://github.com/apache/spark/pull/24947#discussion_r307589185, shall we close this and investigate if we can remove those functions from function registry? That way will cleanly fix this issue.
